### ISSUE TITLE
Xen PVH: Third time lucky

### ISSUE THREAD

**Issue by mato** [mirage/mirage#1159]
After discussing the state of the Xen platform stack with @mirage/core last month, we decided to explore the option of replacing Mini-OS entirely using a _hybrid_ approach with Solo5 providing the low-level "early boot" code and Mirage providing the rest of the stack.

This approach has the advantage of starting from scratch with a minimal and modern code-base, and gets the existing build system infrastructure and freestanding OCaml runtime provided by Solo5 "for free".

I'm happy to report that this is now past PoC stage and the general approach has been shown to work, and I'm opening this issue to track the remaining work in public. This is a replacement for both #998 (Unikraft) and #1007 (update to Mini-OS master).

**Overall status**: Unikernels from `mirage-skeleton` build and run, vchan test unikernels from `ocaml-vchan` likewise. Tested on Xen 4.11, should work on 4.10 or higher.

**Qubes OS status**: ~~Qubes OS 4.0 ships with Xen 4.8 which we do not support. These changes need to be tested against the upcoming Qubes OS 4.1 release (https://github.com/mirage/mirage/issues/1159#issuecomment-656201463).~~ Support for Qubes OS custom Xen 4.8 version is now part of mirage/mirage-xen#23.

**Non-goals**: The existing ARM32 support is removed, this is all x86\_64 only for now. Only Xen PVH v2 is supported, available from Xen 4.10 onwards.

There is a package universe available in `mirage/mirage-dev`, based on **Mirage 3.x** (currently 3.8.1):

    opam repo add mirage-dev-3.x+xen-pvh-via-solo5 git+https://github.com/mirage/mirage-dev.git#3.x+xen-pvh-via-solo5

The following components make up this work. Anything below which does not yet have an active PR is liable to get re-based at any time:

| Component | OPAM Package | Branch | Issue/PR | Released |
| ---- | ---- | ---- | ---- | ---- |
| Solo5 Xen bindings | solo5-bindings-xen | [mato/solo5#xen](https://github.com/mato/solo5/tree/xen) | Solo5/solo5#468 | 0.6.7 :heavy_check_mark: 
| Freestanding runtime | ocaml-freestanding | [mato/ocaml-freestanding#xen](https://github.com/mato/ocaml-freestanding/tree/xen) | mirage/ocaml-freestanding#81 | 0.6.3 :heavy_check_mark: 
| Core platform bindings | mirage-xen | [mato/mirage-xen#xen-pvh-via-solo5](https://github.com/mato/mirage-xen/tree/xen-pvh-via-solo5) | mirage/mirage-xen#23 | 6.0.0 https://github.com/ocaml/opam-repository/pull/17444 :heavy_check_mark: 
| Mirage tool | mirage | [mato/mirage#3.8+xen-pvh-via-solo5](https://github.com/mato/mirage/tree/3.8+xen-pvh-via-solo5) | mirage/mirage#1183 | 3.9.0 https://github.com/ocaml/opam-repository/pull/17485 :heavy_check_mark: 

The following dependencies need API updates, no functional changes intended:

| Component | OPAM Package | Branch | Issue/PR | Released |
| ---- | ---- | ---- | ---- | ---- |
| Block | mirage-block-xen | [mato/mirage-block-xen#xen-pvh-via-solo5](https://github.com/mato/mirage-block-xen/tree/xen-pvh-via-solo5) | mirage/mirage-block-xen#87 | 2.0.0 https://github.com/ocaml/opam-repository/pull/17454 :heavy_check_mark: 
| Bootvar | mirage-bootvar-xen | [mato/mirage-bootvar-xen#xen-pvh-via-solo5](https://github.com/mato/mirage-bootvar-xen/tree/xen-pvh-via-solo5) | mirage/mirage-bootvar-xen#45 | 0.8.0 https://github.com/ocaml/opam-repository/pull/17455 :heavy_check_mark: 
| Console | mirage-console-xen, mirage-console-xen-backend | [mato/mirage-console#xen-pvh-via-solo5](https://github.com/mato/mirage-console/tree/xen-pvh-via-solo5) | mirage/mirage-console#85 | 4.0.0 https://github.com/ocaml/opam-repository/pull/17456 :heavy_check_mark: 
| Crypto | mirage-crypto-pk | OPAM [metadata](https://github.com/mirage/mirage-dev/blob/3.x%2Bxen-pvh-via-solo5/packages/mirage-crypto-pk.0.8.0/opam) change only | | 0.8.6 https://github.com/ocaml/opam-repository/pull/17459 :heavy_check_mark: 
| ~~Io_page~~ | ~~io-page-xen~~ | ~~[mato/io-page#xen-pvh-via-solo5](https://github.com/mato/io-page/tree/xen-pvh-via-solo5)~~ | (no longer required) |
| Network | mirage-net-xen, netchannel | [mato/mirage-net-xen#xen-pvh-via-solo5](https://github.com/mato/mirage-net-xen/tree/xen-pvh-via-solo5) | mirage/mirage-net-xen#99 | 2.0.0 https://github.com/ocaml/opam-repository/pull/17457 :heavy_check_mark: 
| Qubes | mirage-qubes | OPAM [metadata](https://github.com/mirage/mirage-qubes/commit/eda7e8745b988e97b107b6131727f090c5f35dd6) change only | | 0.9.0 https://github.com/ocaml/opam-repository/pull/17471 :heavy_check_mark: 
| Vchan | vchan-xen | [mato/ocaml-vchan#xen-pvh-via-solo5](https://github.com/mato/ocaml-vchan/tree/xen-pvh-via-solo5) | mirage/ocaml-vchan#136 | 6.0.0 https://github.com/ocaml/opam-repository/pull/17458 :heavy_check_mark: 

## Rough TODO list

1. [x] PoC works, Solo5 Xen bindings complete, Mirage "hello, world" works
1. [x] Functionally complete (builds mirage-skeleton, ocaml-vchan tests)
2. [x] Clean up interfaces in mirage-xen (naming of C stubs, ~~`OS` module~~ fold xen-evtchn into mirage-xen)
3. [x] All components above have PRs and review is in progress
4. [x] Wider testing of the new Xen stack, especially from Qubes OS users (see mirage/qubes-mirage-firewall#116)
5. [x] Merge PRs, cut releases and PR to opam-repository, followed by a Mirage 3.x release (see "Released" column in above tables for progress)
6. [x] Forward-port to Mirage 4 (see #1193)

**Comment by avsm** [comment]
Really nice work! Had a quick look at the Bootvar and Io_page API changes, and both look fine. It's lovely to see the MiniOS internals disappear from the bindings.

**Comment by mato** [comment]
Updated to reflect the current status, WIP branches and added a rough TODO list. This is now "almost" functionally complete barring some rough edges which I hope to have sorted soon, after which I will put out a call for wider testing while working on getting the WIP branches into shape for PR and actual review.

**Comment by hannesm** [comment]
I tried your branches / opam repo on my Qubes laptop. It does not work out of the box unfortunately (the start_info is version 0 here, not >= 1 -- Xen version: 4.8.5-14.fc25). The output of a unikernel (I added a log for the command line and moved the `assert(si->version >= 1)` some lines below:

```
Solo5: Xen console: port 0x2, ring @0x00000000FEFFF000
Solo5 Xen start_info->version 0x0
Solo5 Xen cmdline root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH rd.plymouth.enable=0 plymouth.enable=0 rd
Solo5: ABORT: xen/platform.c:142: Assertion `si->version >= 1' failed
```

AFAICT, solo5 platform uses the start_info to figure out the memory size as well, which is not present in version 0 (used by QubesOS). Reading a bit of mini-os, it looks like it's possible to use a different path (https://github.com/mirage/mini-os/blob/48d3b31ce47156d92255a2fe98334dd15a42a097/arch/x86/mm.c#L127). Though, I'm not sure what the timeline of Xen updates in QubesOS are (maybe @marmarek has some comments here?) in order to get a newer start_info struct.

A second issue outlined above is that it looks like QubesOS needs the boot parameter filters, since it's passing something along that won't work with cmdliner/functoria/mirage.

The output of dmesg in the host system:
```
[    0.000000] Linux version 4.19.94-1.pvops.qubes.x86_64 (user@build-fedora4) (gcc version 6.4.1 20170727 (Red Hat 6.4.1-1) (GCC)) #1 SMP Sun Jan 12 05:45:26 UTC 2020
[    0.000000] Command line: root=/dev/mapper/qubes_dom0-root rd.luks.uuid=luks-b7802621-ea7f-4ee1-b51d-7d64100c07d9 rd.lvm.lv=qubes_dom0/root rd.lvm.lv=qubes_dom0/swap i915.alpha_support=1 rhgb quiet rd.qubes.hide_all_usb
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: Enabled xstate features 0x7, context size is 832 bytes, using 'standard' format.
[    0.000000] Released 0 page(s)
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] Xen: [mem 0x0000000000000000-0x0000000000057fff] usable
[    0.000000] Xen: [mem 0x0000000000058000-0x0000000000058fff] reserved
[    0.000000] Xen: [mem 0x0000000000059000-0x000000000009bfff] usable
[    0.000000] Xen: [mem 0x000000000009c000-0x00000000000fffff] reserved
[    0.000000] Xen: [mem 0x0000000000100000-0x000000000fffffff] usable
[    0.000000] Xen: [mem 0x0000000010000000-0x000000001000afff] reserved
[    0.000000] Xen: [mem 0x000000001000b000-0x00000000bf7defff] usable
[    0.000000] Xen: [mem 0x00000000bf7df000-0x00000000bf7fefff] reserved
[    0.000000] Xen: [mem 0x00000000bf7ff000-0x00000000c1216fff] usable
[    0.000000] Xen: [mem 0x00000000c1217000-0x00000000c27cefff] reserved
[    0.000000] Xen: [mem 0x00000000c27cf000-0x00000000c416cfff] usable
[    0.000000] Xen: [mem 0x00000000c416d000-0x00000000c4ae1fff] reserved
[    0.000000] Xen: [mem 0x00000000c4ae2000-0x00000000c4ae3fff] usable
[    0.000000] Xen: [mem 0x00000000c4ae4000-0x00000000c4f8bfff] reserved
[    0.000000] Xen: [mem 0x00000000c4f8c000-0x00000000c5618fff] usable
[    0.000000] Xen: [mem 0x00000000c5619000-0x00000000c563cfff] reserved
[    0.000000] Xen: [mem 0x00000000c563d000-0x00000000c5724fff] usable
[    0.000000] Xen: [mem 0x00000000c5725000-0x00000000c666dfff] reserved
[    0.000000] Xen: [mem 0x00000000c666e000-0x00000000c6680fff] usable
[    0.000000] Xen: [mem 0x00000000c6681000-0x00000000c7cc0fff] reserved
[    0.000000] Xen: [mem 0x00000000c7cc1000-0x00000000cb705fff] usable
[    0.000000] Xen: [mem 0x00000000cb706000-0x00000000ccbfefff] reserved
[    0.000000] Xen: [mem 0x00000000ccbff000-0x00000000ccd7efff] ACPI NVS
[    0.000000] Xen: [mem 0x00000000ccd7f000-0x00000000ccdfefff] ACPI data
[    0.000000] Xen: [mem 0x00000000ccdff000-0x00000000ccdfffff] reserved
[    0.000000] Xen: [mem 0x00000000f80f8000-0x00000000f80f8fff] reserved
[    0.000000] Xen: [mem 0x00000000fec00000-0x00000000fec00fff] reserved
[    0.000000] Xen: [mem 0x00000000fed1c000-0x00000000fed1ffff] reserved
[    0.000000] Xen: [mem 0x00000000fed90000-0x00000000fed91fff] reserved
[    0.000000] Xen: [mem 0x00000000fee00000-0x00000000feefffff] reserved
[    0.000000] Xen: [mem 0x0000000100000000-0x000000013930bfff] usable
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] efi: EFI v2.31 by Lenovo
[    0.000000] efi:  ACPI=0xccdfe000  ACPI 2.0=0xccdfe014  SMBIOS=0xccbfe000  ESRT=0xcc055000 
[    0.000000] SMBIOS 2.7 present.
[    0.000000] DMI: LENOVO 20CLS41500/20CLS41500, BIOS N10ET36W (1.15 ) 06/19/2015
[    0.000000] Hypervisor detected: Xen PV
[    0.045297] tsc: Fast TSC calibration using PIT
[    0.045298] tsc: Detected 2593.966 MHz processor
[    0.045299] tsc: Detected 2593.994 MHz TSC
[    0.045363] e820: update [mem 0x00000000-0x00000fff] usable ==> reserved
[    0.045366] e820: remove [mem 0x000a0000-0x000fffff] usable
[    0.045376] last_pfn = 0x13930c max_arch_pfn = 0x400000000
[    0.045377] Disabled
[    0.045378] x86/PAT: MTRRs disabled, skipping PAT initialization too.
[    0.045381] x86/PAT: Configuration [0-7]: WB  WT  UC- UC  WC  WP  UC  UC  
[    0.045382] last_pfn = 0xcb706 max_arch_pfn = 0x400000000
[    0.045417] Kernel/User page tables isolation: disabled on XEN PV.
[    0.046069] BRK [0x02e01000, 0x02e01fff] PGTABLE
[    0.046070] BRK [0x02e02000, 0x02e02fff] PGTABLE
[    0.047767] BRK [0x02e03000, 0x02e03fff] PGTABLE
[    0.048316] BRK [0x02e04000, 0x02e04fff] PGTABLE
[    0.048864] BRK [0x02e05000, 0x02e05fff] PGTABLE
[    0.049414] BRK [0x02e06000, 0x02e06fff] PGTABLE
[    0.049966] BRK [0x02e07000, 0x02e07fff] PGTABLE
[    0.050514] BRK [0x02e08000, 0x02e08fff] PGTABLE
[    0.051062] BRK [0x02e09000, 0x02e09fff] PGTABLE
[    0.051610] BRK [0x02e0a000, 0x02e0afff] PGTABLE
[    0.052162] BRK [0x02e0b000, 0x02e0bfff] PGTABLE
[    0.052710] BRK [0x02e0c000, 0x02e0cfff] PGTABLE
[    1.153923] Secure boot disabled
[    1.153925] RAMDISK: [mem 0x04000000-0x05694fff]
[    1.153936] ACPI: Early table checksum verification disabled
[    1.153946] ACPI: RSDP 0x00000000CCDFE014 000024 (v02 LENOVO)
[    1.153955] ACPI: XSDT 0x00000000CCDFE1C0 0000F4 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.153989] ACPI: FACP 0x00000000CCDF9000 00010C (v05 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154045] ACPI: DSDT 0x00000000CCDE4000 010E04 (v01 LENOVO TP-N10   00001150 INTL 20120711)
[    1.154058] ACPI: FACS 0x00000000CCD68000 000040
[    1.154071] ACPI: ASF! 0x00000000CCDFD000 0000A5 (v32 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154084] ACPI: HPET 0x00000000CCDFC000 000038 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154097] ACPI: ECDT 0x00000000CCDFB000 000052 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154110] ACPI: APIC 0x00000000CCDF8000 000098 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154123] ACPI: MCFG 0x00000000CCDF7000 00003C (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154136] ACPI: SSDT 0x00000000CCDF6000 000033 (v01 LENOVO TP-SSDT1 00000100 INTL 20120711)
[    1.154149] ACPI: SSDT 0x00000000CCDF5000 000486 (v01 LENOVO TP-SSDT2 00000200 INTL 20120711)
[    1.154162] ACPI: SSDT 0x00000000CCDE3000 0009CB (v01 LENOVO SataAhci 00001000 INTL 20120711)
[    1.154175] ACPI: SSDT 0x00000000CCDE2000 000152 (v01 LENOVO Rmv_Batt 00001000 INTL 20120711)
[    1.154188] ACPI: SSDT 0x00000000CCDE1000 000691 (v01 LENOVO Cpu0Ist  00003000 INTL 20120711)
[    1.154202] ACPI: SSDT 0x00000000CCDE0000 000B74 (v02 LENOVO CpuSsdt  00003000 INTL 20120711)
[    1.154215] ACPI: SSDT 0x00000000CCDDF000 000369 (v02 LENOVO CtdpB    00001000 INTL 20120711)
[    1.154228] ACPI: SSDT 0x00000000CCDDD000 001475 (v01 LENOVO SaSsdt   00003000 INTL 20120711)
[    1.154242] ACPI: SSDT 0x00000000CCDDC000 000394 (v02 LENOVO CppcTabl 00001000 INTL 20120711)
[    1.154257] ACPI: PCCT 0x00000000CCDDB000 00006E (v05 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154270] ACPI: SSDT 0x00000000CCDDA000 000AC4 (v02 LENOVO Cpc_Tabl 00001000 INTL 20120711)
[    1.154283] ACPI: TCPA 0x00000000CCDD9000 000032 (v02 PTL    LENOVO   06040000 LNVO 00000001)
[    1.154298] ACPI: SSDT 0x00000000CCDD8000 0006A3 (v01 Intel_ TpmTable 00001000 INTL 20120711)
[    1.154311] ACPI: UEFI 0x00000000CCDD7000 000042 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154324] ACPI: POAT 0x00000000CCCB2000 000055 (v03 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154338] ACPI: BATB 0x00000000CCDD6000 000046 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154351] ACPI: FPDT 0x00000000CCDD5000 000064 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154364] ACPI: UEFI 0x00000000CCDD4000 0002F6 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154377] ACPI: BGRT 0x00000000CCDD3000 000038 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154390] ACPI: RMAD 0x00000000CCDD2000 0000B0 (v01 LENOVO TP-N10   00001150 PTEC 00000002)
[    1.154426] ACPI: Local APIC address 0xfee00000
[    1.154427] Setting APIC routing to Xen PV.
[    1.154477] NUMA turned off
[    1.154478] Faking a node at [mem 0x0000000000000000-0x000000013930bfff]
[    1.154490] NODE_DATA(0) allocated [mem 0xcaef4000-0xcaf1efff]
[    1.181085] Zone ranges:
[    1.181088]   DMA      [mem 0x0000000000001000-0x0000000000ffffff]
[    1.181089]   DMA32    [mem 0x0000000001000000-0x00000000ffffffff]
[    1.181090]   Normal   [mem 0x0000000100000000-0x000000013930bfff]
[    1.181091]   Device   empty
[    1.181092] Movable zone start for each node
[    1.181095] Early memory node ranges
[    1.181096]   node   0: [mem 0x0000000000001000-0x0000000000057fff]
[    1.181097]   node   0: [mem 0x0000000000059000-0x000000000009bfff]
[    1.181097]   node   0: [mem 0x0000000000100000-0x000000000fffffff]
[    1.181098]   node   0: [mem 0x000000001000b000-0x00000000bf7defff]
[    1.181099]   node   0: [mem 0x00000000bf7ff000-0x00000000c1216fff]
[    1.181099]   node   0: [mem 0x00000000c27cf000-0x00000000c416cfff]
[    1.181100]   node   0: [mem 0x00000000c4ae2000-0x00000000c4ae3fff]
[    1.181100]   node   0: [mem 0x00000000c4f8c000-0x00000000c5618fff]
[    1.181101]   node   0: [mem 0x00000000c563d000-0x00000000c5724fff]
[    1.181102]   node   0: [mem 0x00000000c666e000-0x00000000c6680fff]
[    1.181102]   node   0: [mem 0x00000000c7cc1000-0x00000000cb705fff]
[    1.181103]   node   0: [mem 0x0000000100000000-0x000000013930bfff]
[    1.181105] Reserved but unavailable: 1 pages
[    1.181107] Initmem setup node 0 [mem 0x0000000000001000-0x000000013930bfff]
[    1.181109] On node 0 totalpages: 1048575
[    1.181110]   DMA zone: 64 pages used for memmap
[    1.181110]   DMA zone: 21 pages reserved
[    1.181111]   DMA zone: 3994 pages, LIFO batch:0
[    1.181225]   DMA32 zone: 12662 pages used for memmap
[    1.181226]   DMA32 zone: 810329 pages, LIFO batch:63
[    1.204914]   Normal zone: 3661 pages used for memmap
[    1.204916]   Normal zone: 234252 pages, LIFO batch:63
[    1.211021] p2m virtual area at (____ptrval____), size is a00000
[    1.669313] Remapped 234252 page(s)
[    1.669353] Reserving Intel graphics memory at [mem 0xce000000-0xcfffffff]
[    1.669790] ACPI: PM-Timer IO Port: 0x1808
[    1.669799] ACPI: Local APIC address 0xfee00000
[    1.669836] ACPI: LAPIC_NMI (acpi_id[0x00] high edge lint[0x1])
[    1.669838] ACPI: LAPIC_NMI (acpi_id[0x01] high edge lint[0x1])
[    1.669866] IOAPIC[0]: apic_id 2, version 32, address 0xfec00000, GSI 0-39
[    1.669876] ACPI: INT_SRC_OVR (bus 0 bus_irq 0 global_irq 2 dfl dfl)
[    1.669878] ACPI: INT_SRC_OVR (bus 0 bus_irq 9 global_irq 9 high level)
[    1.669882] ACPI: IRQ0 used by override.
[    1.669884] ACPI: IRQ9 used by override.
[    1.669893] Using ACPI (MADT) for SMP configuration information
[    1.669898] ACPI: HPET id: 0x8086a301 base: 0xfed00000
[    1.669913] smpboot: Allowing 8 CPUs, 4 hotplug CPUs
[    1.669951] [mem 0xd0000000-0xf80f7fff] available for PCI devices
[    1.669955] Booting paravirtualized kernel on Xen
[    1.669955] Xen version: 4.8.5-14.fc25 (preserve-AD)
[    1.669958] clocksource: refined-jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 1910969940391419 ns
[    1.790923] random: get_random_bytes called from start_kernel+0xc3/0x765 with crng_init=0
[    1.790932] setup_percpu: NR_CPUS:64 nr_cpumask_bits:64 nr_cpu_ids:8 nr_node_ids:1
[    1.791212] percpu: Embedded 44 pages/cpu s143360 r8192 d28672 u262144
[    1.791224] pcpu-alloc: s143360 r8192 d28672 u262144 alloc=1*2097152
[    1.791225] pcpu-alloc: [0] 0 1 2 3 4 5 6 7 
[    1.791287] xen: PV spinlocks enabled
[    1.791292] PV qspinlock hash table entries: 256 (order: 0, 4096 bytes)
[    1.791297] Built 1 zonelists, mobility grouping on.  Total pages: 1032167
[    1.791298] Policy zone: Normal
[    1.791300] Kernel command line: root=/dev/mapper/qubes_dom0-root rd.luks.uuid=luks-b7802621-ea7f-4ee1-b51d-7d64100c07d9 rd.lvm.lv=qubes_dom0/root rd.lvm.lv=qubes_dom0/swap i915.alpha_support=1 rhgb quiet rd.qubes.hide_all_usb
[    1.829707] software IO TLB: mapped [mem 0x135000000-0x139000000] (64MB)
[    1.847717] Memory: 3986464K/4194300K available (12300K kernel code, 1486K rwdata, 3884K rodata, 2660K init, 4404K bss, 207836K reserved, 0K cma-reserved)
[    1.847907] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    1.848533] ftrace: allocating 38219 entries in 150 pages
[    1.863942] rcu: Hierarchical RCU implementation.
[    1.863944] rcu: 	RCU restricting CPUs from NR_CPUS=64 to nr_cpu_ids=2.
[    1.863944] 	Tasks RCU enabled.
[    1.863945] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=2
[    1.868303] Using NULL legacy PIC
[    1.868304] NR_IRQS: 4352, nr_irqs: 512, preallocated irqs: 0
[    1.868336] xen:events: Using FIFO-based ABI
[    1.868356] xen: --> pirq=1 -> irq=1 (gsi=1)
[    1.868364] xen: --> pirq=2 -> irq=2 (gsi=2)
[    1.868372] xen: --> pirq=3 -> irq=3 (gsi=3)
[    1.868379] xen: --> pirq=4 -> irq=4 (gsi=4)
[    1.868386] xen: --> pirq=5 -> irq=5 (gsi=5)
[    1.868394] xen: --> pirq=6 -> irq=6 (gsi=6)
[    1.868401] xen: --> pirq=7 -> irq=7 (gsi=7)
[    1.868408] xen: --> pirq=8 -> irq=8 (gsi=8)
[    1.868415] xen: --> pirq=9 -> irq=9 (gsi=9)
[    1.868422] xen: --> pirq=10 -> irq=10 (gsi=10)
[    1.868428] xen: --> pirq=11 -> irq=11 (gsi=11)
[    1.868435] xen: --> pirq=12 -> irq=12 (gsi=12)
[    1.868443] xen: --> pirq=13 -> irq=13 (gsi=13)
[    1.868450] xen: --> pirq=14 -> irq=14 (gsi=14)
[    1.868456] xen: --> pirq=15 -> irq=15 (gsi=15)
[    1.868459] rcu: 	Offload RCU callbacks from CPUs: (none).
[    1.868491] Console: colour dummy device 80x25
[    1.868495] console [tty0] enabled
[    1.868497] console [hvc0] enabled
[    1.868519] ACPI: Core revision 20180810
[    1.875170] clocksource: xen: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    1.875180] Xen: using vcpuop timer interface
[    1.875183] installing Xen timer for CPU 0
[    1.875399] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x256412811b4, max_idle_ns: 440795306987 ns
[    1.875475] Calibrating delay loop (skipped), value calculated using timer frequency.. 5187.98 BogoMIPS (lpj=2593994)
[    1.875477] pid_max: default: 32768 minimum: 301
[    1.875547] Security Framework initialized
[    1.875549] Yama: becoming mindful.
[    1.875562] AppArmor: AppArmor disabled by boot time parameter
[    1.877521] Dentry cache hash table entries: 524288 (order: 10, 4194304 bytes)
[    1.878436] Inode-cache hash table entries: 262144 (order: 9, 2097152 bytes)
[    1.878485] Mount-cache hash table entries: 8192 (order: 4, 65536 bytes)
[    1.878519] Mountpoint-cache hash table entries: 8192 (order: 4, 65536 bytes)
[    1.878977] ENERGY_PERF_BIAS: Set to 'normal', was 'performance'
[    1.878977] ENERGY_PERF_BIAS: View and update with x86_energy_perf_policy(8)
[    1.879022] Last level iTLB entries: 4KB 64, 2MB 8, 4MB 8
[    1.879022] Last level dTLB entries: 4KB 64, 2MB 0, 4MB 0, 1GB 4
[    1.879025] Spectre V1 : Mitigation: usercopy/swapgs barriers and __user pointer sanitization
[    1.879027] Spectre V2 : Mitigation: Full generic retpoline
[    1.879029] Spectre V2 : Spectre v2 / SpectreRSB mitigation: Filling RSB on context switch
[    1.879030] Spectre V2 : Enabling Restricted Speculation for firmware calls
[    1.879071] Spectre V2 : mitigation: Enabling conditional Indirect Branch Prediction Barrier
[    1.879072] Spectre V2 : User space: Mitigation: STIBP via seccomp and prctl
[    1.879073] Speculative Store Bypass: Mitigation: Speculative Store Bypass disabled via prctl and seccomp
[    1.879331] TAA: Mitigation: Clear CPU buffers
[    1.879331] MDS: Mitigation: Clear CPU buffers
[    1.879537] Freeing SMP alternatives memory: 32K
[    1.882500] cpu 0 spinlock event irq 41
[    1.882505] VPMU disabled by hypervisor.
[    1.882696] Performance Events: Broadwell events, PMU not available due to virtualization, using software events only.
[    1.882805] rcu: Hierarchical SRCU implementation.
[    1.883465] random: crng done (trusting CPU's manufacturer)
[    1.883510] NMI watchdog: Perf NMI watchdog permanently disabled
[    1.883602] smp: Bringing up secondary CPUs ...
[    1.883751] installing Xen timer for CPU 1
[    1.884033] cpu 1 spinlock event irq 53
[    1.884033] smp: Brought up 1 node, 2 CPUs
[    1.884033] smpboot: Max logical packages: 4
[    1.884414] devtmpfs: initialized
[    1.884457] x86/mm: Memory block size: 128MB
[    1.884826] PM: Registering ACPI NVS region [mem 0xccbff000-0xccd7efff] (1572864 bytes)
[    1.884826] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 1911260446275000 ns
[    1.884826] futex hash table entries: 512 (order: 3, 32768 bytes)
[    1.884826] pinctrl core: initialized pinctrl subsystem
[    1.884826] RTC time: 14:00:50, date: 07/09/20
[    1.884826] NET: Registered protocol family 16
[    1.884826] xen:grant_table: Grant tables using version 1 layout
[    1.885411] Grant table initialized
[    1.885525] audit: initializing netlink subsys (disabled)
[    1.885536] audit: type=2000 audit(1594303250.162:1): state=initialized audit_enabled=0 res=1
[    1.885536] Detected 1 PCC Subspaces
[    1.885536] Registering PCC driver as Mailbox controller
[    1.885536] ACPI FADT declares the system doesn't support PCIe ASPM, so disable it
[    1.885536] ACPI: bus type PCI registered
[    1.885771] PCI: MMCONFIG for domain 0000 [bus 00-3f] at [mem 0xf8000000-0xfbffffff] (base 0xf8000000)
[    1.885795] PCI: not using MMCONFIG
[    1.885796] PCI: Using configuration type 1 for base access
[    1.887465] cryptd: max_cpu_qlen set to 1000
[    1.887470] ACPI: Added _OSI(Module Device)
[    1.887470] ACPI: Added _OSI(Processor Device)
[    1.887470] ACPI: Added _OSI(3.0 _SCP Extensions)
[    1.887471] ACPI: Added _OSI(Processor Aggregator Device)
[    1.887472] ACPI: Added _OSI(Linux-Dell-Video)
[    1.887473] ACPI: Added _OSI(Linux-Lenovo-NV-HDMI-Audio)
[    1.887477] ACPI: EC: EC started
[    1.887477] ACPI: EC: interrupt blocked
[    1.887487] ACPI: \: Used as first EC
[    1.887489] ACPI: \: GPE=0x25, EC_CMD/EC_SC=0x66, EC_DATA=0x62
[    1.887490] ACPI: \: Used as boot ECDT EC to handle transactions
[    1.906615] ACPI: 12 ACPI AML tables successfully acquired and loaded
[    1.907413] xen: registering gsi 9 triggering 0 polarity 0
[    1.908266] ACPI: [Firmware Bug]: BIOS _OSI(Linux) query ignored
[    1.914653] ACPI Error: Needed type [Reference], found [Integer] 00000000c8f7dda3 (20180810/exresop-69)
[    1.914657] ACPI Error: AE_AML_OPERAND_TYPE, While resolving operands for [OpcodeName unavailable] (20180810/dswexec-427)
[    1.914659] ACPI Error: Method parse/execution failed \_PR.CPU0._PDC, AE_AML_OPERAND_TYPE (20180810/psparse-516)
[    1.915094] ACPI: Dynamic OEM Table Load:
[    1.915101] ACPI: SSDT 0xFFFF8881340F1800 0005AA (v01 PmRef  ApIst    00003000 INTL 20120711)
[    1.915588] ACPI: Dynamic OEM Table Load:
[    1.915594] ACPI: SSDT 0xFFFF88813423EE00 000119 (v02 PmRef  ApCst    00003000 INTL 20120711)
[    1.916019] ACPI: Interpreter enabled
[    1.916044] ACPI: (supports S0 S3 S5)
[    1.916045] ACPI: Using IOAPIC for interrupt routing
[    1.916074] PCI: MMCONFIG for domain 0000 [bus 00-3f] at [mem 0xf8000000-0xfbffffff] (base 0xf8000000)
[    1.916576] PCI: MMCONFIG at [mem 0xf8000000-0xfbffffff] reserved in ACPI motherboard resources
[    1.936202] PCI: Using host bridge windows from ACPI; if necessary, use "pci=nocrs" and report a bug
[    1.936485] ACPI: Enabled 6 GPEs in block 00 to 7F
[    1.939425] ACPI: Power Resource [PUBS] (on)
[    1.939529] acpi PNP0C0A:01: ACPI dock station (docks/bays count: 1)
[    1.939807] ACPI: Power Resource [NVP3] (on)
[    1.939827] ACPI: Power Resource [NVP2] (on)
[    1.943559] ACPI: PCI Interrupt Link [LNKA] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.943640] ACPI: PCI Interrupt Link [LNKB] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.943719] ACPI: PCI Interrupt Link [LNKC] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.943799] ACPI: PCI Interrupt Link [LNKD] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.943876] ACPI: PCI Interrupt Link [LNKE] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.943955] ACPI: PCI Interrupt Link [LNKF] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.944030] ACPI: PCI Interrupt Link [LNKG] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.944108] ACPI: PCI Interrupt Link [LNKH] (IRQs 3 4 5 6 7 9 10 11) *0, disabled.
[    1.944195] ACPI: PCI Root Bridge [PCI0] (domain 0000 [bus 00-3f])
[    1.944198] acpi PNP0A08:00: _OSC: OS supports [ExtendedConfig ASPM ClockPM Segments MSI]
[    1.944338] acpi PNP0A08:00: _OSC: platform does not support [PCIeCapability LTR]
[    1.944407] acpi PNP0A08:00: _OSC: not requesting control; platform does not support [PCIeCapability]
[    1.944408] acpi PNP0A08:00: _OSC: OS requested [PME AER PCIeCapability LTR]
[    1.944409] acpi PNP0A08:00: _OSC: platform willing to grant [PME AER]
[    1.944411] acpi PNP0A08:00: _OSC failed (AE_SUPPORT); disabling ASPM
[    1.944489] PCI host bridge to bus 0000:00
[    1.944491] pci_bus 0000:00: root bus resource [io  0x0000-0x0cf7 window]
[    1.944493] pci_bus 0000:00: root bus resource [io  0x0d00-0xffff window]
[    1.944494] pci_bus 0000:00: root bus resource [mem 0x000a0000-0x000bffff window]
[    1.944495] pci_bus 0000:00: root bus resource [mem 0x000c0000-0x000c3fff window]
[    1.944496] pci_bus 0000:00: root bus resource [mem 0x000c4000-0x000c7fff window]
[    1.944497] pci_bus 0000:00: root bus resource [mem 0x000c8000-0x000cbfff window]
[    1.944498] pci_bus 0000:00: root bus resource [mem 0x000cc000-0x000cffff window]
[    1.944499] pci_bus 0000:00: root bus resource [mem 0x000d0000-0x000d3fff window]
[    1.944500] pci_bus 0000:00: root bus resource [mem 0x000d4000-0x000d7fff window]
[    1.944501] pci_bus 0000:00: root bus resource [mem 0x000d8000-0x000dbfff window]
[    1.944502] pci_bus 0000:00: root bus resource [mem 0x000dc000-0x000dffff window]
[    1.944503] pci_bus 0000:00: root bus resource [mem 0xd0000000-0xfebfffff window]
[    1.944504] pci_bus 0000:00: root bus resource [mem 0xfed40000-0xfed4bfff window]
[    1.944506] pci_bus 0000:00: root bus resource [bus 00-3f]
[    1.944529] pci 0000:00:00.0: [8086:1604] type 00 class 0x060000
[    1.944798] pci 0000:00:02.0: [8086:1616] type 00 class 0x030000
[    1.944863] pci 0000:00:02.0: reg 0x10: [mem 0xf0000000-0xf0ffffff 64bit]
[    1.944896] pci 0000:00:02.0: reg 0x18: [mem 0xe0000000-0xefffffff 64bit pref]
[    1.944917] pci 0000:00:02.0: reg 0x20: [io  0x3000-0x303f]
[    1.944970] pci 0000:00:02.0: BAR 2: assigned to efifb
[    1.945170] pci 0000:00:03.0: [8086:160c] type 00 class 0x040300
[    1.945223] pci 0000:00:03.0: reg 0x10: [mem 0xf1230000-0xf1233fff 64bit]
[    1.945569] pci 0000:00:14.0: [8086:9cb1] type 00 class 0x0c0330
[    1.945638] pci 0000:00:14.0: reg 0x10: [mem 0xf1220000-0xf122ffff 64bit]
[    1.945845] pci 0000:00:14.0: PME# supported from D3hot D3cold
[    1.945965] pci 0000:00:16.0: [8086:9cba] type 00 class 0x078000
[    1.946037] pci 0000:00:16.0: reg 0x10: [mem 0xf1239000-0xf123901f 64bit]
[    1.946254] pci 0000:00:16.0: PME# supported from D0 D3hot D3cold
[    1.946396] pci 0000:00:19.0: [8086:15a2] type 00 class 0x020000
[    1.946458] pci 0000:00:19.0: reg 0x10: [mem 0xf1200000-0xf121ffff]
[    1.946482] pci 0000:00:19.0: reg 0x14: [mem 0xf123e000-0xf123efff]
[    1.946506] pci 0000:00:19.0: reg 0x18: [io  0x3080-0x309f]
[    1.946704] pci 0000:00:19.0: PME# supported from D0 D3hot D3cold
[    1.946831] pci 0000:00:1b.0: [8086:9ca0] type 00 class 0x040300
[    1.946900] pci 0000:00:1b.0: reg 0x10: [mem 0xf1234000-0xf1237fff 64bit]
[    1.947106] pci 0000:00:1b.0: PME# supported from D0 D3hot D3cold
[    1.947221] pci 0000:00:1c.0: [8086:9c9a] type 01 class 0x060400
[    1.947472] pci 0000:00:1c.0: PME# supported from D0 D3hot D3cold
[    1.947521] pci 0000:00:1c.0: Enabling MPC IRBNCE
[    1.947525] pci 0000:00:1c.0: Intel PCH root port ACS workaround enabled
[    1.947685] pci 0000:00:1c.1: [8086:9c94] type 01 class 0x060400
[    1.947934] pci 0000:00:1c.1: PME# supported from D0 D3hot D3cold
[    1.947981] pci 0000:00:1c.1: Enabling MPC IRBNCE
[    1.947985] pci 0000:00:1c.1: Intel PCH root port ACS workaround enabled
[    1.948159] pci 0000:00:1d.0: [8086:9ca6] type 00 class 0x0c0320
[    1.948224] pci 0000:00:1d.0: reg 0x10: [mem 0xf123d000-0xf123d3ff]
[    1.948482] pci 0000:00:1d.0: PME# supported from D0 D3hot D3cold
[    1.948613] pci 0000:00:1f.0: [8086:9cc3] type 00 class 0x060100
[    1.949013] pci 0000:00:1f.2: [8086:9c83] type 00 class 0x010601
[    1.949074] pci 0000:00:1f.2: reg 0x10: [io  0x30a8-0x30af]
[    1.949097] pci 0000:00:1f.2: reg 0x14: [io  0x30b4-0x30b7]
[    1.949120] pci 0000:00:1f.2: reg 0x18: [io  0x30a0-0x30a7]
[    1.949143] pci 0000:00:1f.2: reg 0x1c: [io  0x30b0-0x30b3]
[    1.949166] pci 0000:00:1f.2: reg 0x20: [io  0x3060-0x307f]
[    1.949190] pci 0000:00:1f.2: reg 0x24: [mem 0xf123c000-0xf123c7ff]
[    1.949317] pci 0000:00:1f.2: PME# supported from D3hot
[    1.949424] pci 0000:00:1f.3: [8086:9ca2] type 00 class 0x0c0500
[    1.949483] pci 0000:00:1f.3: reg 0x10: [mem 0xf1238000-0xf12380ff 64bit]
[    1.949551] pci 0000:00:1f.3: reg 0x20: [io  0xefa0-0xefbf]
[    1.949718] pci 0000:00:1f.6: [8086:9ca4] type 00 class 0x118000
[    1.949801] pci 0000:00:1f.6: reg 0x10: [mem 0xf123b000-0xf123bfff 64bit]
[    1.950263] pci 0000:02:00.0: [10ec:5227] type 00 class 0xff0000
[    1.950325] pci 0000:02:00.0: reg 0x10: [mem 0xf1100000-0xf1100fff]
[    1.950657] pci 0000:02:00.0: supports D1 D2
[    1.950658] pci 0000:02:00.0: PME# supported from D1 D2 D3hot D3cold
[    1.950883] pci 0000:00:1c.0: PCI bridge to [bus 02]
[    1.950897] pci 0000:00:1c.0:   bridge window [mem 0xf1100000-0xf11fffff]
[    1.951046] pci 0000:03:00.0: [8086:095b] type 00 class 0x028000
[    1.951143] pci 0000:03:00.0: reg 0x10: [mem 0xf1000000-0xf1001fff 64bit]
[    1.951512] pci 0000:03:00.0: PME# supported from D0 D3hot D3cold
[    1.951767] pci 0000:00:1c.1: PCI bridge to [bus 03]
[    1.951781] pci 0000:00:1c.1:   bridge window [mem 0xf1000000-0xf10fffff]
[    1.951971] xen: registering gsi 13 triggering 1 polarity 0
[    1.953079] ACPI: EC: interrupt unblocked
[    1.953099] ACPI: EC: event unblocked
[    1.953110] ACPI: \_SB_.PCI0.LPC_.EC__: GPE=0x25, EC_CMD/EC_SC=0x66, EC_DATA=0x62
[    1.953112] ACPI: \_SB_.PCI0.LPC_.EC__: Used as boot DSDT EC to handle transactions and events
[    1.953138] xen:balloon: Initialising balloon driver
[    1.953209] pci 0000:00:02.0: vgaarb: VGA device added: decodes=io+mem,owns=mem,locks=none
[    1.953209] pci 0000:00:02.0: vgaarb: bridge control possible
[    1.953209] pci 0000:00:02.0: vgaarb: setting as boot device
[    1.953209] vgaarb: loaded
[    1.953209] SCSI subsystem initialized
[    1.953422] libata version 3.00 loaded.
[    1.953443] ACPI: bus type USB registered
[    1.953457] usbcore: registered new interface driver usbfs
[    1.953464] usbcore: registered new interface driver hub
[    1.953475] usbcore: registered new device driver usb
[    1.953498] pps_core: LinuxPPS API ver. 1 registered
[    1.953499] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    1.953503] PTP clock support registered
[    1.953550] EDAC MC: Ver: 3.0.0
[    1.953595] Registered efivars operations
[    1.953595] PCI: Using ACPI for IRQ routing
[    1.959700] PCI: pci_cache_line_size set to 64 bytes
[    1.959817] e820: reserve RAM buffer [mem 0x00058000-0x0005ffff]
[    1.959818] e820: reserve RAM buffer [mem 0x0009c000-0x0009ffff]
[    1.959819] e820: reserve RAM buffer [mem 0xbf7df000-0xbfffffff]
[    1.959820] e820: reserve RAM buffer [mem 0xc1217000-0xc3ffffff]
[    1.959821] e820: reserve RAM buffer [mem 0xc416d000-0xc7ffffff]
[    1.959822] e820: reserve RAM buffer [mem 0xc4ae4000-0xc7ffffff]
[    1.959823] e820: reserve RAM buffer [mem 0xc5619000-0xc7ffffff]
[    1.959824] e820: reserve RAM buffer [mem 0xc5725000-0xc7ffffff]
[    1.959825] e820: reserve RAM buffer [mem 0xc6681000-0xc7ffffff]
[    1.959826] e820: reserve RAM buffer [mem 0xcb706000-0xcbffffff]
[    1.959827] e820: reserve RAM buffer [mem 0x13930c000-0x13bffffff]
[    1.959921] NetLabel: Initializing
[    1.959921] NetLabel:  domain hash size = 128
[    1.959922] NetLabel:  protocols = UNLABELED CIPSOv4 CALIPSO
[    1.959934] NetLabel:  unlabeled traffic allowed by default
[    1.959944] clocksource: Switched to clocksource tsc-early
[    1.969549] VFS: Disk quotas dquot_6.6.0
[    1.969563] VFS: Dquot-cache hash table entries: 512 (order 0, 4096 bytes)
[    1.969583] hugetlbfs: disabling because there are no supported hugepage sizes
[    1.969600] pnp: PnP ACPI init
[    1.970089] system 00:00: [mem 0x00000000-0x0009ffff] could not be reserved
[    1.970091] system 00:00: [mem 0x000e0000-0x000e3fff] could not be reserved
[    1.970092] system 00:00: [mem 0x000e4000-0x000e7fff] could not be reserved
[    1.970093] system 00:00: [mem 0x000e8000-0x000ebfff] could not be reserved
[    1.970095] system 00:00: [mem 0x000ec000-0x000effff] could not be reserved
[    1.970096] system 00:00: [mem 0x000f0000-0x000fffff] could not be reserved
[    1.970097] system 00:00: [mem 0x00100000-0xcfffffff] could not be reserved
[    1.970098] system 00:00: [mem 0xfec00000-0xfed3ffff] could not be reserved
[    1.970100] system 00:00: [mem 0xfed4c000-0xffffffff] could not be reserved
[    1.970104] system 00:00: Plug and Play ACPI device, IDs PNP0c01 (active)
[    1.970200] pnp 00:01: [Firmware Bug]: PNP resource [mem 0xfed10000-0xfed13fff] covers only part of 0000:00:00.0 Intel MCH; extending to [mem 0xfed10000-0xfed17fff]
[    1.970215] system 00:01: [io  0x1800-0x189f] has been reserved
[    1.970217] system 00:01: [io  0x0800-0x087f] has been reserved
[    1.970218] system 00:01: [io  0x0880-0x08ff] has been reserved
[    1.970219] system 00:01: [io  0x0900-0x097f] has been reserved
[    1.970221] system 00:01: [io  0x0980-0x09ff] has been reserved
[    1.970222] system 00:01: [io  0x0a00-0x0a7f] has been reserved
[    1.970223] system 00:01: [io  0x0a80-0x0aff] has been reserved
[    1.970224] system 00:01: [io  0x0b00-0x0b7f] has been reserved
[    1.970225] system 00:01: [io  0x0b80-0x0bff] has been reserved
[    1.970227] system 00:01: [io  0x15e0-0x15ef] has been reserved
[    1.970228] system 00:01: [io  0x1600-0x167f] has been reserved
[    1.970229] system 00:01: [io  0x1640-0x165f] has been reserved
[    1.970231] system 00:01: [mem 0xf8000000-0xfbffffff] could not be reserved
[    1.970232] system 00:01: [mem 0xfed1c000-0xfed1ffff] has been reserved
[    1.970234] system 00:01: [mem 0xfed10000-0xfed17fff] has been reserved
[    1.970235] system 00:01: [mem 0xfed18000-0xfed18fff] has been reserved
[    1.970236] system 00:01: [mem 0xfed19000-0xfed19fff] has been reserved
[    1.970237] system 00:01: [mem 0xfed45000-0xfed4bfff] has been reserved
[    1.970240] system 00:01: Plug and Play ACPI device, IDs PNP0c02 (active)
[    1.970283] xen: registering gsi 8 triggering 1 polarity 0
[    1.970314] pnp 00:02: Plug and Play ACPI device, IDs PNP0b00 (active)
[    1.970320] xen: registering gsi 1 triggering 1 polarity 0
[    1.970343] pnp 00:03: Plug and Play ACPI device, IDs LEN0071 PNP0303 (active)
[    1.970348] xen: registering gsi 12 triggering 1 polarity 0
[    1.970375] pnp 00:04: Plug and Play ACPI device, IDs LEN0046 PNP0f13 (active)
[    1.971518] pnp 00:05: Plug and Play ACPI device, IDs SMO1200 PNP0c31 (active)
[    1.971716] system 00:06: [mem 0xd0010000-0xd001ffff] has been reserved
[    1.971717] system 00:06: [mem 0xd0000000-0xd000ffff] has been reserved
[    1.971720] system 00:06: Plug and Play ACPI device, IDs PNP0c02 (active)
[    1.971749] pnp: PnP ACPI: found 7 devices
[    1.986530] PM-Timer failed consistency check  (0xffffff) - aborting.
[    1.986600] pci 0000:00:1c.0: PCI bridge to [bus 02]
[    1.986612] pci 0000:00:1c.0:   bridge window [mem 0xf1100000-0xf11fffff]
[    1.986632] pci 0000:00:1c.1: PCI bridge to [bus 03]
[    1.986643] pci 0000:00:1c.1:   bridge window [mem 0xf1000000-0xf10fffff]
[    1.986666] pci_bus 0000:00: resource 4 [io  0x0000-0x0cf7 window]
[    1.986667] pci_bus 0000:00: resource 5 [io  0x0d00-0xffff window]
[    1.986668] pci_bus 0000:00: resource 6 [mem 0x000a0000-0x000bffff window]
[    1.986669] pci_bus 0000:00: resource 7 [mem 0x000c0000-0x000c3fff window]
[    1.986670] pci_bus 0000:00: resource 8 [mem 0x000c4000-0x000c7fff window]
[    1.986671] pci_bus 0000:00: resource 9 [mem 0x000c8000-0x000cbfff window]
[    1.986672] pci_bus 0000:00: resource 10 [mem 0x000cc000-0x000cffff window]
[    1.986673] pci_bus 0000:00: resource 11 [mem 0x000d0000-0x000d3fff window]
[    1.986674] pci_bus 0000:00: resource 12 [mem 0x000d4000-0x000d7fff window]
[    1.986675] pci_bus 0000:00: resource 13 [mem 0x000d8000-0x000dbfff window]
[    1.986676] pci_bus 0000:00: resource 14 [mem 0x000dc000-0x000dffff window]
[    1.986677] pci_bus 0000:00: resource 15 [mem 0xd0000000-0xfebfffff window]
[    1.986679] pci_bus 0000:00: resource 16 [mem 0xfed40000-0xfed4bfff window]
[    1.986680] pci_bus 0000:02: resource 1 [mem 0xf1100000-0xf11fffff]
[    1.986681] pci_bus 0000:03: resource 1 [mem 0xf1000000-0xf10fffff]
[    1.986760] NET: Registered protocol family 2
[    1.986907] tcp_listen_portaddr_hash hash table entries: 2048 (order: 3, 32768 bytes)
[    1.987018] TCP established hash table entries: 32768 (order: 6, 262144 bytes)
[    1.987210] TCP bind hash table entries: 32768 (order: 7, 524288 bytes)
[    1.987286] TCP: Hash tables configured (established 32768 bind 32768)
[    1.987328] UDP hash table entries: 2048 (order: 4, 65536 bytes)
[    1.987357] UDP-Lite hash table entries: 2048 (order: 4, 65536 bytes)
[    1.987395] NET: Registered protocol family 1
[    1.987400] NET: Registered protocol family 44
[    1.987437] pci 0000:00:02.0: Video device with shadowed ROM at [mem 0x000c0000-0x000dffff]
[    1.987508] xen: registering gsi 16 triggering 0 polarity 1
[    1.987526] xen: --> pirq=16 -> irq=16 (gsi=16)
[    1.987759] xen: registering gsi 23 triggering 0 polarity 1
[    1.987768] xen: --> pirq=23 -> irq=23 (gsi=23)
[    1.987994] PCI: CLS 64 bytes, default 64
[    1.988024] Unpacking initramfs...
[    2.375473] Freeing initrd memory: 23124K
[    2.376597] Initialise system trusted keyrings
[    2.376604] Key type blacklist registered
[    2.376667] workingset: timestamp_bits=36 max_order=20 bucket_order=0
[    2.377754] zbud: loaded
[    2.783683] alg: No test for 842 (842-generic)
[    2.783733] alg: No test for 842 (842-scomp)
[    2.796600] NET: Registered protocol family 38
[    2.796603] Key type asymmetric registered
[    2.796604] Asymmetric key parser 'x509' registered
[    2.796616] Block layer SCSI generic (bsg) driver version 0.4 loaded (major 244)
[    2.796666] io scheduler noop registered
[    2.796667] io scheduler deadline registered
[    2.796702] io scheduler cfq registered (default)
[    2.796703] io scheduler mq-deadline registered
[    2.796772] atomic64_test: passed for x86-64 platform with CX8 and with SSE
[    2.796960] xen: registering gsi 17 triggering 0 polarity 1
[    2.796977] xen: --> pirq=17 -> irq=17 (gsi=17)
[    2.797163] xen: registering gsi 18 triggering 0 polarity 1
[    2.797176] xen: --> pirq=18 -> irq=18 (gsi=18)
[    2.797321] efifb: probing for efifb
[    2.797324] efi: EFI_MEMMAP is not enabled.
[    2.799764] efifb: No BGRT, not showing boot graphics
[    2.799765] efifb: framebuffer at 0xe0000000, using 8100k, total 8100k
[    2.799766] efifb: mode is 1920x1080x32, linelength=7680, pages=1
[    2.799766] efifb: scrolling: redraw
[    2.799768] efifb: Truecolor: size=8:8:8:8, shift=24:16:8:0
[    2.799832] fbcon: Deferring console take-over
[    2.799833] fb0: EFI VGA frame buffer device
[    2.799841] intel_idle: MWAIT substates: 0x11142120
[    2.799842] intel_idle: v0.4.1 model 0x3D
[    2.799844] intel_idle: intel_idle yielding to none
[    2.800274] ACPI: AC Adapter [AC] (on-line)
[    2.800325] input: Lid Switch as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0C0D:00/input/input0
[    2.800349] ACPI: Lid Switch [LID]
[    2.800389] input: Sleep Button as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0C0E:00/input/input1
[    2.800421] ACPI: Sleep Button [SLPB]
[    2.800467] input: Power Button as /devices/LNXSYSTM:00/LNXPWRBN:00/input/input2
[    2.800501] ACPI: Power Button [PWRF]
[    2.800698] ACPI BIOS Error (bug): Could not resolve [\_PR.CPU0._CST], AE_NOT_FOUND (20180810/psargs-330)
[    2.800702] ACPI Error: Method parse/execution failed \_PR.CPU1._CST, AE_NOT_FOUND (20180810/psparse-516)
[    2.800760] Warning: Processor Platform Limit not supported.
[    2.803780] thermal LNXTHERM:00: registered as thermal_zone0
[    2.803781] ACPI: Thermal Zone [THM0] (68 C)
[    2.809358] Serial: 8250/16550 driver, 32 ports, IRQ sharing enabled
[    2.819061] hpet_acpi_add: no address or irqs in _CRS
[    2.819156] Non-volatile memory driver v1.3
[    2.819308] Linux agpgart interface v0.103
[    2.822466] tpm_tis 00:05: 1.2 TPM (device-id 0x0, rev-id 78)
[    2.830971] battery: ACPI: Battery Slot [BAT0] (battery present)
[    2.857644] battery: ACPI: Battery Slot [BAT1] (battery present)
[    2.859461] tpm tpm0: TPM is disabled/deactivated (0x6)
[    2.911847] ahci 0000:00:1f.2: version 3.0
[    2.911911] xen: registering gsi 19 triggering 0 polarity 1
[    2.911926] xen: --> pirq=19 -> irq=19 (gsi=19)
[    2.911970] ahci 0000:00:1f.2: BIOS update required for Link Power Management support
[    2.922163] ahci 0000:00:1f.2: AHCI 0001.0300 32 slots 3 ports 6 Gbps 0x3 impl SATA mode
[    2.922165] ahci 0000:00:1f.2: flags: 64bit ncq pm led clo only pio slum part deso sadm sds apst 
[    2.922613] scsi host0: ahci
[    2.922895] scsi host1: ahci
[    2.923135] scsi host2: ahci
[    2.923172] ata1: SATA max UDMA/133 abar m2048@0xf123c000 port 0xf123c100 irq 57
[    2.923173] ata2: SATA max UDMA/133 abar m2048@0xf123c000 port 0xf123c180 irq 57
[    2.923174] ata3: DUMMY
[    2.923248] libphy: Fixed MDIO Bus: probed
[    2.923347] usbcore: registered new interface driver usbserial_generic
[    2.923351] usbserial: USB Serial support registered for generic
[    2.923377] i8042: PNP: PS/2 Controller [PNP0303:KBD,PNP0f13:MOU] at 0x60,0x64 irq 1,12
[    2.925534] serio: i8042 KBD port at 0x60,0x64 irq 1
[    2.925538] serio: i8042 AUX port at 0x60,0x64 irq 12
[    2.925620] mousedev: PS/2 mouse device common for all mice
[    2.925780] rtc_cmos 00:02: RTC can wake from S4
[    2.926050] rtc_cmos 00:02: registered as rtc0
[    2.926052] rtc_cmos 00:02: alarms up to one month, y3k, 114 bytes nvram
[    2.926098] device-mapper: uevent: version 1.0.3
[    2.926163] device-mapper: ioctl: 4.39.0-ioctl (2018-04-03) initialised: dm-devel@redhat.com
[    2.926599] hidraw: raw HID events driver (C) Jiri Kosina
[    2.926625] usbcore: registered new interface driver usbhid
[    2.926626] usbhid: USB HID core driver
[    2.926684] drop_monitor: Initializing network drop monitor service
[    2.927019] Initializing XFRM netlink socket
[    2.927115] NET: Registered protocol family 10
[    2.928105] input: AT Translated Set 2 keyboard as /devices/platform/i8042/serio0/input/input3
[    2.933275] Segment Routing with IPv6
[    2.933299] mip6: Mobile IPv6
[    2.933301] NET: Registered protocol family 17
[    2.933536] mce: Using 2 MCE banks
[    2.933550] RAS: Correctable Errors collector initialized.
[    2.933559] AVX2 version of gcm_enc/dec engaged.
[    2.933559] AES CTR mode by8 optimization enabled
[    2.965229] sched_clock: Marking stable (2958279411, 6939767)->(2966574246, -1355068)
[    2.965735] registered taskstats version 1
[    2.965745] Loading compiled-in X.509 certificates
[    2.965761] zswap: loaded using pool lzo/zbud
[    2.976512] Key type big_key registered
[    2.981573] Key type encrypted registered
[    2.981585] ima: Allocated hash algorithm: sha1
[    2.989476] ima: Error Communicating to TPM chip
[    2.996709] ima: Error Communicating to TPM chip
[    3.003030] ima: Error Communicating to TPM chip
[    3.011413] ima: Error Communicating to TPM chip
[    3.018751] ima: Error Communicating to TPM chip
[    3.026103] ima: Error Communicating to TPM chip
[    3.032411] ima: Error Communicating to TPM chip
[    3.039794] ima: Error Communicating to TPM chip
[    3.048724]   Magic number: 12:128:30
[    3.048865] rtc_cmos 00:02: setting system clock to 2020-07-09 14:00:51 UTC (1594303251)
[    3.234153] ata2: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[    3.234167] ata1: SATA link up 6.0 Gbps (SStatus 133 SControl 300)
[    3.235046] ata2.00: LPM support broken, forcing max_power
[    3.235190] ata2.00: ACPI cmd ef/02:00:00:00:00:a0 (SET FEATURES) succeeded
[    3.235192] ata2.00: ACPI cmd f5/00:00:00:00:00:a0 (SECURITY FREEZE LOCK) filtered out
[    3.235244] ata2.00: ACPI cmd ef/10:09:00:00:00:a0 (SET FEATURES) succeeded
[    3.235473] ata2.00: ATA-9: TS256GMTS400, N1126KB, max UDMA/133
[    3.235475] ata2.00: 500118192 sectors, multi 1: LBA48 NCQ (depth 32), AA
[    3.235838] ata2.00: LPM support broken, forcing max_power
[    3.235943] ata2.00: ACPI cmd ef/02:00:00:00:00:a0 (SET FEATURES) succeeded
[    3.235945] ata2.00: ACPI cmd f5/00:00:00:00:00:a0 (SECURITY FREEZE LOCK) filtered out
[    3.235972] ata2.00: ACPI cmd ef/10:09:00:00:00:a0 (SET FEATURES) succeeded
[    3.236264] ata2.00: configured for UDMA/133
[    3.236960] ata1.00: LPM support broken, forcing max_power
[    3.237117] ata1.00: ACPI cmd ef/02:00:00:00:00:a0 (SET FEATURES) succeeded
[    3.237119] ata1.00: ACPI cmd f5/00:00:00:00:00:a0 (SECURITY FREEZE LOCK) filtered out
[    3.237179] ata1.00: ACPI cmd ef/10:09:00:00:00:a0 (SET FEATURES) succeeded
[    3.237409] ata1.00: ATA-9: LITEON LCH-512V2S, FC92201, max UDMA/133
[    3.237410] ata1.00: 1000215216 sectors, multi 2: LBA48 NCQ (depth 32), AA
[    3.237719] ata1.00: LPM support broken, forcing max_power
[    3.237855] ata1.00: ACPI cmd ef/02:00:00:00:00:a0 (SET FEATURES) succeeded
[    3.237856] ata1.00: ACPI cmd f5/00:00:00:00:00:a0 (SECURITY FREEZE LOCK) filtered out
[    3.237889] ata1.00: ACPI cmd ef/10:09:00:00:00:a0 (SET FEATURES) succeeded
[    3.238190] ata1.00: configured for UDMA/133
[    3.248664] scsi 0:0:0:0: Direct-Access     ATA      LITEON LCH-512V2 201  PQ: 0 ANSI: 5
[    3.248924] sd 0:0:0:0: Attached scsi generic sg0 type 0
[    3.249005] sd 0:0:0:0: [sda] 1000215216 512-byte logical blocks: (512 GB/477 GiB)
[    3.249018] sd 0:0:0:0: [sda] Write Protect is off
[    3.249019] sd 0:0:0:0: [sda] Mode Sense: 00 3a 00 00
[    3.249042] sd 0:0:0:0: [sda] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[    3.249062] scsi 1:0:0:0: Direct-Access     ATA      TS256GMTS400     6KB  PQ: 0 ANSI: 5
[    3.249193] sd 1:0:0:0: Attached scsi generic sg1 type 0
[    3.249301] sd 1:0:0:0: [sdb] 500118192 512-byte logical blocks: (256 GB/238 GiB)
[    3.249314] sd 1:0:0:0: [sdb] Write Protect is off
[    3.249315] sd 1:0:0:0: [sdb] Mode Sense: 00 3a 00 00
[    3.249337] sd 1:0:0:0: [sdb] Write cache: enabled, read cache: enabled, doesn't support DPO or FUA
[    3.249957] sd 0:0:0:0: [sda] Attached SCSI disk
[    3.250780]  sdb: sdb1 sdb2 sdb3
[    3.251221] sd 1:0:0:0: [sdb] Attached SCSI disk
[    3.252807] Freeing unused decrypted memory: 2040K
[    3.254183] Freeing unused kernel image memory: 2660K
[    3.257446] Write protecting the kernel read-only data: 18432k
[    3.270254] Freeing unused kernel image memory: 2024K
[    3.270418] Freeing unused kernel image memory: 212K
[    3.270427] rodata_test: all tests were successful
[    3.270429] Run /init as init process
[    3.375708] systemd[1]: systemd 231 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN)
[    3.375837] systemd[1]: Detected virtualization xen.
[    3.375847] systemd[1]: Detected architecture x86-64.
[    3.375853] systemd[1]: Running in initial RAM disk.
[    3.375885] systemd[1]: Set hostname to <dom0>.
[    3.379424] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x256412811b4, max_idle_ns: 440795306987 ns
[    3.379671] clocksource: Switched to clocksource tsc
[    3.476439] systemd[1]: Reached target Swap.
[    3.477137] systemd[1]: Created slice System Slice.
[    3.477152] systemd[1]: Reached target Timers.
[    3.477163] systemd[1]: Reached target Local File Systems.
[    3.477300] systemd[1]: Created slice system-systemd\x2dcryptsetup.slice.
[    3.551815] xen:xen_evtchn: Event-channel device installed
[    3.554076] audit: type=1130 audit(1594303251.984:2): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-journald comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.554079] audit: type=1130 audit(1594303251.985:3): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=kmod-static-nodes comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.555735] audit: type=1130 audit(1594303252.004:4): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-tmpfiles-setup comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.555737] audit: type=1130 audit(1594303252.005:5): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-tmpfiles-setup-dev comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.567608] xen_pciback: backend is vpci
[    3.598839] fbcon: Taking over console
[    3.617711] Console: switching to colour frame buffer device 240x67
[    3.646812] audit: type=1130 audit(1594303252.048:6): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-modules-load comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=failed'
[    3.646816] audit: type=1130 audit(1594303252.059:7): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-sysctl comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.652758] audit: type=1130 audit(1594303252.098:8): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-vconsole-setup comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    3.684482] pciback 0000:00:19.0: seizing device
[    3.684617] xen: registering gsi 20 triggering 0 polarity 1
[    3.684641] xen: --> pirq=20 -> irq=20 (gsi=20)
[    3.787876] pciback 0000:03:00.0: seizing device
[    3.787990] xen: registering gsi 17 triggering 0 polarity 1
[    3.787994] Already setup the GSI :17
[    3.828857] psmouse serio1: synaptics: queried max coordinates: x [..5712], y [..4780]
[    3.859492] psmouse serio1: synaptics: queried min coordinates: x [1232..], y [1074..]
[    3.859499] psmouse serio1: synaptics: Trying to set up SMBus access
[    3.862141] psmouse serio1: synaptics: SMbus companion is not ready yet
[    3.891948] pciback 0000:00:14.0: seizing device
[    3.892079] xen: registering gsi 16 triggering 0 polarity 1
[    3.892083] Already setup the GSI :16
[    3.892431] pciback 0000:00:1d.0: seizing device
[    3.892539] xen: registering gsi 23 triggering 0 polarity 1
[    3.892542] Already setup the GSI :23
[    3.922060] psmouse serio1: synaptics: Touchpad model: 1, fw: 8.1, id: 0x1e2b1, caps: 0xf002a3/0x943300/0x12e800/0x10000, board id: 3075, fw id: 2560
[    3.922068] psmouse serio1: synaptics: serio: Synaptics pass-through port at isa0060/serio1/input0
[    3.960123] input: SynPS/2 Synaptics TouchPad as /devices/platform/i8042/serio1/input/input5
[    4.163352] audit: type=1130 audit(1594303252.612:9): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=dracut-cmdline comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    4.175682] audit: type=1130 audit(1594303252.622:10): pid=1 uid=0 auid=4294967295 ses=4294967295 msg='unit=systemd-udevd comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    4.427222] xen: registering gsi 16 triggering 0 polarity 1
[    4.427228] Already setup the GSI :16
[    4.436155] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    4.436440] ehci-pci: EHCI PCI platform driver
[    4.473717] Setting dangerous option alpha_support - tainting kernel
[    4.475664] i915 0000:00:02.0: enabling device (0006 -> 0007)
[    4.475745] xen: registering gsi 16 triggering 0 polarity 1
[    4.475750] Already setup the GSI :16
[    4.479584] checking generic (e0000000 7e9000) vs hw (e0000000 10000000)
[    4.479585] fb: switching to inteldrmfb from EFI VGA
[    4.479626] Console: switching to colour dummy device 80x25
[    4.483223] [drm] Replacing VGA console driver
[    4.588253] [drm] Supports vblank timestamp caching Rev 2 (21.10.2013).
[    4.588255] [drm] Driver supports precise vblank timestamp query.
[    4.588328] i915 0000:00:02.0: vgaarb: changed VGA decodes: olddecodes=io+mem,decodes=io+mem:owns=mem
[    4.658893] [drm] Initialized i915 1.6.0 20180719 for 0000:00:02.0 on minor 0
[    4.659477] ACPI: Video Device [VID] (multi-head: yes  rom: no  post: no)
[    4.659871] input: Video Bus as /devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A08:00/LNXVIDEO:00/input/input7
[    4.824060] fbcon: inteldrmfb (fb0) is primary device
[    4.824123] Console: switching to colour frame buffer device 240x67
[    4.828016] i915 0000:00:02.0: fb0: inteldrmfb frame buffer device
[    5.129952] psmouse serio2: trackpoint: IBM TrackPoint firmware: 0x0e, buttons: 3/3
[    5.325263] input: TPPS/2 IBM TrackPoint as /devices/platform/i8042/serio1/serio2/input/input6
```

**Comment by marmarek** [comment]
> A second issue outlined above is that it looks like QubesOS needs the boot parameter filters, since it's passing something along that won't work with cmdliner/functoria/mirage.

Already available: `qvm-features VMNAME no-default-kernelopts 1`
Then, `kernelopts` property is the only thing that gets passed to the unikernel and you can set it freely to any value.

> Though, I'm not sure what the timeline of Xen updates in QubesOS are (maybe @marmarek has some comments here?)

Qubes 4.1 should be out in few months (rc1 probably next month, then 2-3 months until final release). This will have Xen 4.13.
But Qubes 4.0 will be there for a while, so if supporting older Xen too isn't too complex, I think it's worth adding it.

**Comment by mato** [comment]
@marmarek 

> Qubes 4.1 should be out in few months (rc1 probably next month, then 2-3 months until final release). This will have Xen 4.13.
> But Qubes 4.0 will be there for a while, so if supporting older Xen too isn't too complex, I think it's worth adding it.

The goal of this work was to eliminate/replace as much legacy code as possible, with prejudice. Therefore, I wrote it to target the Xen PVHv2 ABIs only. Given that Xen only lists PVHv2 as "fully supported" from 4.10 onwards, I would strongly prefer not to try and support earlier versions.

**Comment by marmarek** [comment]
Fair enough.

**Comment by hannesm** [comment]
hi @marmarek ,

> `qvm-features VMNAME no-default-kernelopts 1`

thanks, that works nicely.

> `kernelopts` 

NB this is part of `qvm-prefs` (not `qvm-features`).

> 4.1rc likely next month

great, I'm looking forward to test this work on qubes with a more recent Xen. I'd be fine to test an early beta of 4.1, if there's documentation for how to upgrade from a Qubes 4.0 (or a binary usb stick release so I can reinstall my QubesOS). please let us know when there's such an upgrade path or beta release.

**Comment by mato** [comment]
All components now have PRs ready for review; I've updated the description here with the current status.

**Comment by mato** [comment]
See #1184 for discussion / work in progress on supporting the Xen version shipped by the existing Qubes OS 4 release.

**Comment by mato** [comment]
Status update: Qubes OS 4.0 support is now done (Solo5/solo5#480 and added to mirage/mirage-xen#23). Most of the discussion is in mirage/qubes-mirage-firewall#116. Further fixes to importing grant mappings from other Xen domains have landed in mirage/mirage-xen#23.

I have updated the package universe (see description) to Mirage 3.8.1 and to include the above; you should be able to build Mirage/Xen unikernels for both Xen 4.10+ _and_ Qubes OS 4+ from it with no problems.

**Comment by hannesm** [comment]
Update: the dependencies have been merged into opam-repository, and mirage 3.9.0 is tagged and PRed to opam-repository. PR #1193 is a forward-port of the changes to master. We plan to announce the release on mailing lists etc. once mirage 3.9.0 is merged into opam-repository.

**Comment by mato** [comment]
All done here!
